src/value.cpp

```diff
@@ -0,0 +1,417 @@
+MIGRAPHX_VISIT_VALUE_TYPES(MIGRAPHX_VALUE_GENERATE_DEFINE_METHODS)
+
+bool value::is_array() const { return x ? x->get_type() == array_type : false; }
+const std::vector<value>& value::value::get_array() const
+{
+    auto* r = this->if_array();
+    assert(r);
+    return *r;
+}
+const std::vector<value>* value::if_array() const { return x ? x->if_array() : nullptr; }
+
+bool value::is_object() const { return x ? x->get_type() == object_type : false; }
+const std::vector<value>& value::get_object() const
+{
+    auto* r = this->if_object();
+    assert(r);
+    return *r;
+}
+const std::vector<value>* value::if_object() const { return x ? x->if_array() : nullptr; }
+
+bool value::is_null() const { return x == nullptr; }
+
+const std::string& value::get_key() const { return key; }
+
+std::vector<value>* if_array_impl(const std::shared_ptr<value_base_impl>& x)
+{
+    if(!x)
+        return nullptr;
+    return x->if_array();
+}
+
+std::vector<value>& get_array_impl(const std::shared_ptr<value_base_impl>& x)
+{
+    auto* a = if_array_impl(x);
+    assert(a);
+    return *a;
+}
+
+value* find_impl(const std::shared_ptr<value_base_impl>& x, const std::string& key)
+{
+    auto* a = if_array_impl(x);
+    if(a == nullptr)
+        return nullptr;
+    auto* lookup = x->if_object();
+    if(lookup == nullptr)
+        return nullptr;
+    auto it = lookup->find(key);
+    if(it == lookup->end())
+        return a->data() + a->size();
+    return std::addressof((*a)[it->second]);
+}
+
+value* value::find(const std::string& pkey) { return find_impl(x, pkey); }
+
+const value* value::find(const std::string& pkey) const { return find_impl(x, pkey); }
+bool value::contains(const std::string& pkey) const
+{
+    auto it = find(pkey);
+    if(it == nullptr)
+        return false;
+    if(it == end())
+        return false;
+    return true;
+}
+std::size_t value::size() const
+{
+    auto* a = if_array_impl(x);
+    if(a == nullptr)
+        return 0;
+    return a->size();
+}
+bool value::empty() const { return size() == 0; }
+const value* value::data() const
+{
+    auto* a = if_array_impl(x);
+    if(a == nullptr)
+        return nullptr;
+    return a->data();
+}
+value* value::data()
+{
+    auto* a = if_array_impl(x);
+    if(a == nullptr)
+        return nullptr;
+    return a->data();
+}
+value* value::begin()
+{
+    // cppcheck-suppress assertWithSideEffect
+    assert(data());
+    return data();
+}
+const value* value::begin() const
+{
+    assert(data());
+    return data();
+}
+value* value::end() { return begin() + size(); }
+const value* value::end() const { return begin() + size(); }
+
+value& value::front() { return *begin(); }
+const value& value::front() const { return *begin(); }
+value& value::back() { return *std::prev(end()); }
+const value& value::back() const { return *std::prev(end()); }
+value& value::at(std::size_t i)
+{
+    auto* a = if_array_impl(x);
+    if(a == nullptr)
```

Comment:
for several of these functions, they're (mostly) doing the same error checking and throwing the same message. Is it worthwhile to create simple error functions that can be reused?